plugins/contents/fps_contents/fileid.py

```diff
@@ -0,0 +1,175 @@
+
+    def __aiter__(self):
+        return self
+
+    async def __anext__(self):
+        await self._event.wait()
+        self._event.clear()
+        return self._change
+
+    def notify(self, change):
+        self._change = change
+        self._event.set()
+
+
+class FileIdManager:
+
+    db_path: str
+    initialized: asyncio.Event
+    watchers: Dict[str, List[Watcher]]
+
+    def __init__(self, db_path: str = "fileid.db"):
+        self.db_path = db_path
+        self.initialized = asyncio.Event()
+        self.watchers = {}
+        asyncio.create_task(self.watch_files())
+
+    async def get_id(self, path: str) -> Optional[str]:
+        await self.initialized.wait()
+        async with aiosqlite.connect(self.db_path) as db:
+            async with db.execute("SELECT id FROM fileids WHERE path = ?", (path,)) as cursor:
+                async for idx, in cursor:
+                    return idx
+                return None
+
+    async def get_path(self, idx: str) -> Optional[str]:
+        await self.initialized.wait()
+        async with aiosqlite.connect(self.db_path) as db:
+            async with db.execute("SELECT path FROM fileids WHERE id = ?", (idx,)) as cursor:
+                async for path, in cursor:
+                    return path
+                return None
+
+    async def index(self, path: str) -> Optional[str]:
+        await self.initialized.wait()
+        async with aiosqlite.connect(self.db_path) as db:
+            apath = AsyncPath(path)
+            if not await apath.exists():
+                return None
+
+            idx = uuid4().hex
+            mtime = (await apath.stat()).st_mtime
+            await db.execute("INSERT INTO fileids VALUES (?, ?, ?)", (idx, path, mtime))
+            await db.commit()
+            return idx
+
+    async def watch_files(self):
+        async with aiosqlite.connect(self.db_path) as db:
+            await db.execute(
+                "CREATE TABLE IF NOT EXISTS fileids "
+                "(id TEXT PRIMARY KEY, path TEXT NOT NULL, mtime REAL NOT NULL)"
+            )
+            await db.commit()
+
+        async with aiosqlite.connect(self.db_path) as db:
+            async for path in AsyncPath().rglob("*"):
+                idx = uuid4().hex
+                mtime = (await path.stat()).st_mtime
+                await db.execute("INSERT INTO fileids VALUES (?, ?, ?)", (idx, str(path), mtime))
+            await db.commit()
+            self.initialized.set()
+
+            async for changes in awatch("."):
+                deleted_paths = []
+                added_paths = []
+                for change, changed_path in changes:
+                    # get relative path
+                    changed_path = str(Path(changed_path).relative_to(Path().absolute()))
+
+                    if change == Change.deleted:
+                        async with db.execute(
+                            "SELECT * FROM fileids WHERE path = ?", (changed_path,)
```

Comment:
It's generally not a good idea to select all columns when you know what you need.  In this case, you really only need an existence check.  For SQLite, this can probably be:
```suggestion
                            "SELECT COUNT(1) FROM fileids WHERE path = ?", (changed_path,)
```